primitives/src/lib.rs

```diff
@@ -61,6 +61,12 @@ pub type Nonce = u32;
 /// A hash of some data used by the chain.
 pub type Hash = sp_core::H256;
 
+/// Can be used to represent an infinite computation. One can use it together
+/// with [Result<Bottom, T>](Result) to indicate that a computation can finish
+/// only when some error occurs, otherwise it will spin forever
+/// (compiler/type-checker will enforce this property).
```

Comment:
This is kinda cool (⊥ is my favourite type in general), but I'm not convinced it's worth the effort if it's not a native type. I doubt everyone will start using it now, so we will only have this in a couple semi-random places, where it'll confuse any readers.

If we want to adopt this, then I would suggest making an ADR for that and actually requiring people to put ⊥s everywhere where there should be a never ending computation.

Reply:
Can't we use https://doc.rust-lang.org/std/primitive.never.html ?

Reply:
https://doc.rust-lang.org/reference/types/never.html

Reply:
Right -- the question is then if we need `Bottom` at all

Reply:
I am ok with introducing it just in few random places (why do we need to use it everywhere immediately?), is more descriptive for infinite loops/services than `Result<(), _>`, it gives some free additional compiler checks, it doesn't require any additional maintenance (no methods, converters etc.), and can be switched to `!` when it become stable (are we going to use it then or pretend it's not there?). ADR for it is an overkill, since it will be a rust's feature anyway. I can change it's name for something more descriptive. It was interesting to see that you can do it in rust. Anyway, should I remove it?

Reply:
The problem with doing it in just a couple random places is that there is additional mental load when trying to figure out why some places have it and others don't – and when the answer is "because it was done in one specific PR" it's not even something that can really be guessed. There is value in consistency. This would be slightly less of a problem if we used native Rust functionalities, but still somewhat, we don't use all the Rust functionalities that exist and would technically improve compiler checks. See for example the infamous lifetimed keychain reference in ABFT.

That being said, I see a couple reasonable approaches to this:
1. Leave this PR in its current shape, but create an ADR saying that we use `Bottom` for all functions where it's applicable and switch to `!` as soon as it's stable. This way we get temporarily confusing code, that will eventually get homogenized, which is fine. The ADR can even be very short, the argument for using this type is pretty straightforward.
2. Remove `Bottom`, but create an ADR for using some type like it eventually, probably once `!` is stabilized.
3. Remove `Bottom` and create an ADR explicitly forbidding the use of `!` to avoid confusion.
4. Remove `Bottom` and do nothing.
This is roughly in my order of preference. Note that 4 is essentially doing 3 implicitly, as there will be a discussion like this every time someone tries to re-add `Bottom`. For 2 & 3 you don't have to create any ADR immediately, it's just something that would eventually have to be done. Also, even if we go for 4 and then accept a PR with `Bottom` (probably in the form of a stabilized `!`), we would still have to at least plan an effort to homogenize the code.

tl;dr Either leave `Bottom` in and create an ADR, or remove bottom and optionally create a task for writing an ADR.